### PR TITLE
7.8.x patch 1

### PR DESCRIPTION
**Pull request by matthewrmshin**
Cherry picked: #2916, #2937, #2941, #2943, #2944.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
#2916, #2937 probably aren't needed for a bugfix release though should be harmless so fine.

**Comment by oliver-sanders** [comment]
Can we get #2938 in?

**Comment by matthewrmshin** [comment]
Yes. We can raise another PR on the 7.8.x branch.

**Comment by sadielbartholomew** [comment]
> Can we get #2938 in?

>  > Yes. We can raise another PR on the 7.8.x branch.

I'm reviewing that PR now to escalate this.

**Comment by hjoliver** [comment]
Also:
- [ ] #2942

File: bin/cylc-cycle-point [hunk 1]
```diff
@@ -266,9 +264,4 @@ def main():
 
 
 if __name__ == "__main__":
-    try:
-        main()
-    except Exception as exc:
-        if cylc.flags.debug:
-            raise
-        sys.exit(str(exc))
+    main()
```
Comment (hjoliver) [inline bin/cylc-cycle-point]: I guess I didn't review the original PR to master, but why this change? (My old "users should see an error message, not a traceback" argument is probably wrong, but we should be consistent, or remove this from all the commands, no?)

Comment (matthewrmshin) [reply bin/cylc-cycle-point]: No, your original assumption was not wrong. However:
* Time has changed and many users are now familiar enough with Python that the trace back is no longer scary - and it helps us and them when they send us bug reports.
* Catching known exception here is good, but catching *all* generic exceptions here is not so good - especially when you have those cryptic system errors - which is much better being a trace back.

You are right. I should changed everything.

Comment (hjoliver) [reply bin/cylc-cycle-point]: Yes, I'm happy to have this removed from all commands.

Comment (oliver-sanders) [reply bin/cylc-cycle-point]: ##### Word of warning: In Python3 when you catch an exception its context is not forgotten which results in multi-part traceback which gets really messy really quick.

For example a pattern we use a lot in Cylc is to catch a nasty `OSError` or `KeyError` or whatever and raise a more user-friendly `CylcException`, `ConfigError` or whatnot. This pattern is at Python3 fundamentally broken as you will end up raising *both* the original exception *as well* as the new one.

For example say we are trying to catch `ZeroDivisionError` and hide it from the user replacing it with a more informative `Exception`...

```python
>>> try:
...     1/0
... except ZeroDivisionError:
...     # raise a completely different exception
...     raise Exception('foo')
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ZeroDivisionError: division by zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
Exception: foo
```

**Comment by hjoliver** [comment]
For future changes that need to target 7.8.x and 8.0, IMO we should raise duplicate separate PRs rather than a combo cherry-pick branch - easier to keep track of what's been merged. Or ask the PR author to do that. (I think this particular combo PR was necessary for historical reasons).

**Comment by hjoliver** [comment]
@kinow - can you do a quick sanity check and merge on this (tomorrow); if @oliver-sanders doesn't overnight.

**Comment by kinow** [comment]
Spare time after Spanish classes, so reviewed this one @hjoliver . Leave to @oliver-sanders do the final check & merge :+1:

**Comment by hjoliver** [comment]
> Would have liked to include bugfix #2938 .

That fix can still go in as another PR - no need to block this merge, is there?

**Comment by matthewrmshin** [comment]
This is patch 1. There can be many more.
